### PR TITLE
Made animation loop interruptible and disposable.

### LINKED ISSUE #13151: Inescapable and non-disposable animation loop
##### Description of the problem

The `WebGLRenderer` has an internal `loop` function as part of the `animate` function, looking at the code it will loop indefinitely and there is no hook for disposal either:

	function loop( time ) {
		if ( onAnimationFrame !== null ) onAnimationFrame( time );

		var device = vr.getDevice();

		if ( device && device.isPresenting ) {
			device.requestAnimationFrame( loop );
		} else {
			window.requestAnimationFrame( loop );
		}
	}

(Looked into this after my `animate` callback was still being executed after calling `renderer.dispose()`, which seems like it should not be the case.)

##### Three.js version

- [x] Dev
- [x] r88

##### Browser

- [x] All of them
- [ ] Chrome
- [ ] Firefox
- [ ] Internet Explorer

##### OS

- [x] All of them
- [ ] Windows
- [ ] macOS
- [ ] Linux
- [ ] Android
- [ ] iOS

### PR DESCRIPTION
**Pull request by brunnerh**
Dispose now should cancel the animation and calling `animate(null)` should do so as well.

Fixes #13151

### CODE REVIEW

**Comment by brunnerh** [comment]
Missed that `isAnimating` has to be reset when breaking the loop via `animate(null)`, otherwise the next call to animate will not work as expected...

**Comment by Mugen87** [comment]
I still think it's better to stop the animation loop with `cancelAnimationFrame()` since it's exactly intended for this purpose.

**Comment by brunnerh** [comment]
In theory i would agree, but here this would complicate the code. `requestAnimationFrame` is called either on the window or the device depending on state. To properly cancel this we then would have to also keep track of where the frame was requested.

You are of course free to realize this via cancelling; i am not particularly attached to the method by which this is resolved.

**Comment by mrdoob** [comment]
I find that logic confusing... is this what you're trying to do?

```js
if ( onAnimationFrame === null ) isAnimating = false;
if ( isAnimating === false ) return;
```

**Comment by brunnerh** [comment]
That would be a cleaner way of writing that, yes. The reset of isAnimating was patched in later, after i noticed that it is necessary.

**Comment by brunnerh** [comment]
Adjusted the code and removed another guard expression that now appears to be unnecessary.

**Comment by mrdoob** [comment]
Another approach:

```js
function stop() {

	isAnimating = false;

}

this.animate = function ( callback ) {

	onAnimationFrame = callback;
	onAnimationFrame !== null ? start() : stop();
 
};
```

**Comment by brunnerh** [comment]
That seems sensible. I added that in and also refactored some duplicate code that was in both `start` and `loop`.

Currently in `this.dispose` there is `isAnimating = false` which now could be replaced with `stop()`. I am wondering about the naming, though. `isAnimating = false` seems a lot more descriptive than `stop()`. How about changing the function names to `startAnimation` and `stopAnimation` (or `...Animating`)?

File: src/renderers/WebGLRenderer.js [hunk 1]
```diff
@@ -1050,6 +1050,22 @@ function WebGLRenderer( parameters ) {
...
 		if ( isAnimating ) return;
 
+		requestLoopFrame();
+
+		isAnimating = true;
+
+	}
+
+	function stop()
+	{
```
Comment (mrdoob) [inline src/renderers/WebGLRenderer.js]: Can you put curly braces in the same line?  😇

**Comment by mrdoob** [comment]
> How about changing the function names to `startAnimation` and `stopAnimation` 

Sounds good 👍

**Comment by brunnerh** [comment]
Renamed all related functions, adjusted `dispose` and fixed the braces.

I feel like it would have been more effective for you to just have rejected this and transfer/adjust the code yourself :thinking:.

**Comment by mrdoob** [comment]
> I feel like it would have been more effective for you to just have rejected this and transfer/adjust the code yourself :thinking:

Probably. But then you would have lost the credit 🤔

**Comment by mrdoob** [comment]
Looks good now!

**Comment by mrdoob** [comment]
Thanks!

**Comment by brunnerh** [comment]
> Probably. But then you would have lost the credit :thinking:

:man_shrugging:

> Thanks!

:+1:

**Pull request merged**